docs/graph_engine/temporal.rst

```diff
@@ -0,0 +1,144 @@
+
+Temporal Nodes
+--------------
+
+Temporal nodes are nodes that can be added or removed from a graph at different times. The JSON schema for temporal nodes extends the schema for `regular<data_spec.html>`_ nodes with two new fields: created_at and removed_at, with values recorded in `Unix time <https://en.wikipedia.org/wiki/Unix_time>`_ format. The following example demonstrates a temporal node in JSON format:
+
+.. code-block:: json
+
+	{
+		"node_id": 42,
+		"node_type": 0,
+		"node_weight": 1.0,
+		"binary_feature": {"0": "microsoft", "1": "xbox"},
+		"created_at": 13,
+		"removed_at": 23,
+	}
+
+This example describes a node with an ID of 42, two binary features, and a single outgoing edge. The node exists from time 13 to time 23. If the created_at field is not specified, it defaults to 0. If the removed_at field is not specified, the node is assumed to be always present after it was created. DeepGNN does not yet support changing node types over time. A workaround is to delete the node and create a new node with a new type or store the node type as a feature.
+
+Temporal Edges
+--------------
+
+Temporal edges are edges that can be added or removed from the graph at specific times. Edges have two fields: created_at and removed_at, with values recorded in Unix time format. The following example extends the previous example to include temporal edges:
+
+.. code-block:: json
+
+	{
+		"node_id": 42,
+		"node_type": 0,
+		"node_weight": 1.0,
+		"created_at": 13,
+		"removed_at": 23,
+		"edge": [
+			{
+				"src_id": 42,
+				"dst_id": 99,
+				"edge_type": 0,
+				"weight": 2.0,
+				"created_at": 13,
+				"removed_at": 15
+			},
+			{
+				"src_id": 42,
+				"dst_id": 99,
+				"edge_type": 0,
+				"weight": 2.0,
+				"created_at": 17
+			},
+			{
+				"src_id": 42,
+				"dst_id": 101,
+				"edge_type": 1,
+				"weight": 1.0,
+				"removed_at": 19
+			}
+		]
+	}
+
+The node with ID 42 has two outgoing edges with types 0 and 1. The first edge exists during intervals [13, 15) and [17, 23), while the second edge is created with the node at time 13 and removed at 19. If the created_at field is not specified, it defaults to the created_at of the node. Similarly, the default value for removed_at is the same as the removed_at of the node.
+
+Temporal Features
+-----------------
+
+Temporal features are node or edge features that can change values over time. To support this behavior, the schema for existing features is expanded. For binary and dense features, the format includes a union of two types:
+
+A list of values for features that are present during the node/edge lifetime.
+A list of objects with three fields: values, created_at, and removed_at.
+The default values for created_at and removed_at are similar to those for nodes and edges. Sparse features are extended in a similar way, with the accepted types being a union of a sparse feature object and a list of such objects. The following example demonstrates the use of temporal features:
```

Comment:
"The default values for created_at and removed_at are similar to those for nodes and edges." Should just say what the default values are.